pyzo/core/shell.py

```diff
@@ -728,12 +744,17 @@ def _handleCarriageReturn(self, text):
         # Remove last line if it ended with CR
         cursor = self._cursor1
         if (self._lastline_had_cr and not text.startswith("\n")) or (
-            text.startswith("\r") and not text[1:].startswith("\n")
+            text.startswith("\r")
+            and not text[1:].startswith("\n")
+            and not self._lastline_had_lf
         ):
-            cursor.movePosition(cursor.PreviousBlock, cursor.KeepAnchor, 1)
+            cursor.movePosition(cursor.StartOfLine, cursor.KeepAnchor, 1)
             cursor.removeSelectedText()
         # Is this new line ending in CR?
         self._lastline_had_cr = text.endswith("\r")
+        # Is this new line ending in LF?
+        self._lastline_had_lf = text.endswith("\n")
```

Comment:
Could you please init this attribute where we also init `self._lastline_had_cr`?